app/services/bulk_upload/lettings/year2023/row_parser.rb

```diff
@@ -732,45 +749,45 @@ def validate_nulls
+    if location.scheme != scheme && location_field.present?
       block_log_creation!
-      errors.add(:field_17, "Scheme code must relate to a location that is owned by owning organisation or managing organisation", category: :setup)
+      errors.add(location_field, "#{scheme_or_management_group.capitalize} code must relate to a #{location_or_scheme} that is owned by owning organisation or managing organisation", category: :setup)
     end
   end
 
   def validate_location_exists
-    if scheme && field_17.present? && location.nil?
-      errors.add(:field_17, "Location could not be found with the provided scheme code", category: :setup)
+    if scheme && location_id.present? && location.nil? && location_field.present?
+      errors.add(location_field, "#{location_or_scheme.capitalize} could not be found with the provided #{scheme_or_management_group} code", category: :setup)
     end
   end
 
   def validate_location_data_given
-    if supported_housing? && field_17.blank?
-      errors.add(:field_17, I18n.t("validations.not_answered", question: "scheme code"), category: "setup")
+    if supported_housing? && location_id.blank? && location_field.present?
+      errors.add(location_field, I18n.t("validations.not_answered", question: "#{location_or_scheme} code"), category: "setup")
     end
   end
 
   def validate_scheme_related
-    return unless field_16.present? && scheme.present?
+    return unless scheme_id.present? && scheme.present?
 
     owned_by_owning_org = owning_organisation && scheme.owning_organisation == owning_organisation
     owned_by_managing_org = managing_organisation && scheme.owning_organisation == managing_organisation
 
-    unless owned_by_owning_org || owned_by_managing_org
+    if !(owned_by_owning_org || owned_by_managing_org) && scheme_field.present?
       block_log_creation!
-      errors.add(:field_16, "This management group code does not belong to your organisation, or any of your stock owners / managing agents", category: :setup)
+      errors.add(scheme_field, "This #{scheme_or_management_group} code does not belong to your organisation, or any of your stock owners / managing agents", category: :setup)
     end
   end
 
   def validate_scheme_exists
-    if field_16.present? && scheme.nil?
-      errors.add(:field_16, "The management group code is not correct", category: :setup)
+    if scheme_id.present? && scheme_field.present? && scheme.nil?
+      errors.add(scheme_field, "The #{scheme_or_management_group} code is not correct", category: :setup)
     end
   end
 
   def validate_scheme_data_given
-    if supported_housing? && field_16.blank?
-      errors.add(:field_16, I18n.t("validations.not_answered", question: "management group code"), category: "setup")
+    if supported_housing? && scheme_field.present? && scheme_id.blank?
+      errors.add(scheme_field, I18n.t("validations.not_answered", question: "#{scheme_or_management_group} code"), category: "setup")
     end
```

Comment:
can scheme_field/location_field be nil?

Reply:
What if they haven't entered any of the location/scheme values. Would scheme_field.present? be false?

Reply:
in that case we don't know which is the scheme field or location field so they'll be nil

Reply:
everywhere we're adding errors to scheme_field or location_field I've checked it exists first, and if it is nil another error is added to field_16 saying it must be answered, which there is a test for

Reply:
okay, gotcha! tyty!